packages/xsnap/test/xsnap-eof.test.js

```diff
@@ -0,0 +1,200 @@
+  const fromXsnap = xsnapProcess.stdio[4];
+  const stderrP = text(xsnapProcess.stderr);
+
+  xsnapProcess.on('error', err => {
+    exitedPKit.reject(err);
+  });
+
+  xsnapProcess.on('exit', (code, signal) => {
+    exitedPKit.resolve({ code, signal });
+  });
+
+  return { vat, xsnapProcess, toXsnap, fromXsnap, stderrP, exitedPKit };
+}
+
+async function issueCommandAndWait(worker) {
+  let issueCommandError;
+  let vatCloseError;
+  const { vat, stderrP, exitedPKit } = worker;
+
+  try {
+    // this will error out since we are intentionally
+    // breaking communications with the worker
+    await vat.issueStringCommand('0');
+  } catch (err) {
+    issueCommandError = err;
+  }
+
+  // wait for the worker to exit
+  const [strErr, { code: vatExitCode, signal: vatExitSignal }] =
+    await Promise.all([stderrP, exitedPKit.promise]);
+
+  try {
+    // this will error out since vat has died.
+    await vat.close();
+  } catch (err) {
+    vatCloseError = err;
+  }
+  return {
+    issueCommandError,
+    vatCloseError,
+    strErr,
+    vatExitCode,
+    vatExitSignal,
+  };
+}
+
+test('xsnap-worker complains while trying to READ an answer when pipes are closed', async t => {
+  const worker = await spawnReflectiveWorker(
+    async function handleCommand(_message) {
+      // the worker is blocked on read here, so we should close "fromXsnap" pipe first
+      worker.fromXsnap.end();
+      worker.fromXsnap.destroy();
+      worker.toXsnap.end();
+      worker.toXsnap.destroy();
+      return new Uint8Array();
+    },
+  );
+
+  const {
+    issueCommandError,
+    vatCloseError,
+    strErr,
+    vatExitCode,
+    vatExitSignal,
+  } = await issueCommandAndWait(worker);
+
+  t.not(issueCommandError, undefined, 'issueCommand() must produce and error');
+  t.not(vatCloseError, 'vat.close() must produce and error');
+  t.is(vatExitSignal, null, 'vat exit signal must be null');
+  t.is(vatExitCode, 2 /* E_IO_ERROR */, 'vat exit code must be 2');
+  t.is(
+    strErr,
+    'Got EOF on netstring read. Has parent died?\n',
+    'stderr must be "Got EOF on netstring read. Has parent died?"',
+  );
+});
+
+test('xsnap-worker complains while trying to WRITE when pipes are closed', async t => {
+  const worker = await spawnReflectiveWorker(
+    async function handleCommand(_message) {
+      // the worker is blocked on read here, so we should close "fromXsnap" pipe first
+      worker.fromXsnap.end();
+      worker.fromXsnap.destroy();
+      // write a fake but valid response into a pipe buffer,
+      // forcing the worker to move to the next state where it would
+      // attempt to write out return value from its own "handleCommand()"
+      // into a pipe which we've just closed above.
+      await makeNetstringWriter(makeNodeWriter(worker.toXsnap)).next([
+        QUERY_RESPONSE_BUF,
+        new Uint8Array(),
+      ]);
+      worker.toXsnap.end();
+      worker.toXsnap.destroy();
+      return new Uint8Array();
```

Comment:
actually I have an alternative idea based on how I/O works in Node.js:

```suggestion
      // Close the writer stream after the response has been sent
      // forcing the worker to move to the next state where it would
      // attempt to write out return value from its own "handleCommand()"
      // into a pipe which we've just closed above.
      setImmediate(() => {
        worker.toXsnap.end();
        worker.toXsnap.destroy();
      });
      return new Uint8Array();
```

Reply:
If we abandon any pretense that we are simulating parent being killed and simply focus on individual pipe failures, then we can do even better:

```
  async function onRequest(worker, _message) {
    // the worker is blocked on read here, so we close "fromXsnap" pipe
    // so the worker will get IO error (SIGPIPE) when it moves to next state after we return from here 
    worker.fromXsnap.end();
    worker.fromXsnap.destroy();
    return new Uint8Array();
  }
```

Reply:
Right, closing `toXsnap` is not strictly necessary in this test, but I guess it doesn't hurt.

Reply:
With new tests we have two cases where we test closure of all pipes and three where we test individual pipe failures.